File: companions/common/elliott_companion/smoke.py
from __future__ import annotations

import argparse
import hashlib
import json
import time
import urllib.error
import urllib.request
from pathlib import Path
from typing import Any


def _request(url: str, value: dict[str, Any] | None = None) -> Any:
    data = None if value is None else json.dumps(value).encode("utf-8")
    request = urllib.request.Request(
        url,
        data=data,
        method="GET" if data is None else "POST",
        headers={"content-type": "application/json"},
    )
    with urllib.request.urlopen(request, timeout=10) as response:
        return json.loads(response.read())

def _baseline_payload(comparison: dict[str, Any]) -> dict[str, Any]:
    payload = json.loads(json.dumps(comparison))
    for name in (
        "candidate",
        "candidateSnapshotId",
        "benchmarkGates",
        "footprintLimits",
        "confidenceLevel",
        "bootstrapIterations",
        "multipleComparisonCount",
        "requiredConstraints",
    ):
        payload.pop(name)
    payload["operation"] = "baseline"
    payload["run"]["state"] = {
        "_tag": "dataset-ready",
        "datasetId": payload["dataset"]["id"],
        "datasetDigest": payload["dataset"]["digest"],
    }
    payload["run"].pop("optimizationSeed", None)
    payload["targetFootprintBytes"] = 128
    payload.pop("evaluationPlanDigest")
    canonical = json.dumps(
        payload,
        ensure_ascii=False,
        separators=(",", ":"),
        sort_keys=True,
    )
    payload["evaluationPlanDigest"] = (
        f"sha256:{hashlib.sha256(canonical.encode('utf-8')).hexdigest()}"
    )
    return payload


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--endpoint", required=True)
    parser.add_argument("--path", required=True)
    parser.add_argument("--request", type=Path, required=True)
    parser.add_argument(
        "--kind",
        choices=["optimizer", "benchmark", "evaluator", "baseline"],
        required=True,
    )
    args = parser.parse_args()
    for _ in range(100):
        try:
            health = _request(args.endpoint.rstrip("/") + "/healthz")
            if health == {"status": "ok"}:
                break
        except urllib.error.URLError:
            time.sleep(0.05)
    else:
        raise SystemExit("companion did not become healthy")
    payload = json.loads(args.request.read_text(encoding="utf-8"))
    if args.kind == "baseline":
        payload = _baseline_payload(payload)
    result = _request(args.endpoint.rstrip("/") + args.path, payload)
    if args.kind == "optimizer":
        if result.get("runId") != payload["run"]["id"]:
            raise SystemExit("optimizer smoke result runId mismatch")
        if result.get("paused") is not False or len(result.get("candidates", [])) != 1:
            raise SystemExit(
                "optimizer smoke result did not contain one completed candidate"
            )
        candidate = result["candidates"][0]
        if candidate.get("runId") != payload["run"]["id"]:
            raise SystemExit("candidate runId mismatch")
        if not str(candidate.get("candidateDigest", "")).startswith("sha256:"):
            raise SystemExit("candidate digest is missing")
    elif args.kind == "benchmark":
        if result.get("benchmarkRef") != payload["benchmarkRef"]:
            raise SystemExit("benchmark smoke result reference mismatch")
        if result.get("status") != "passed" or result.get("passed") is not True:
            raise SystemExit("benchmark smoke result did not pass")
        if not str(result.get("reportDigest", "")).startswith("sha256:"):
            raise SystemExit("benchmark report digest is missing")
    elif args.kind == "evaluator":
        if result.get("runId") != payload["run"]["id"]:
            raise SystemExit("evaluation smoke result runId mismatch")
        if result.get("candidateId") != payload["candidate"]["id"]:
            raise SystemExit("evaluation smoke result candidateId mismatch")
        if result.get("passed") is not True:
            raise SystemExit("evaluation smoke result did not pass")
        if len(result.get("benchmarks", [])) != len(payload["benchmarkGates"]):
            raise SystemExit("evaluation smoke result omitted benchmark gates")
    else:
        if result.get("runId") != payload["run"]["id"]:
            raise SystemExit("baseline smoke result runId mismatch")
        if {
            item.get("split") for item in result.get("caseResults", [])
        } != {"validation", "holdout"}:
            raise SystemExit("baseline smoke did not isolate evaluation splits")
        if len(result.get("trajectoryDigests", [])) != len(
            result.get("caseResults", [])
        ):
            raise SystemExit("baseline smoke result omitted trajectories")
        if {
            item.get("category") for item in result.get("footprints", [])
        } != {"prompt", "inference", "runtime"}:
            raise SystemExit("baseline smoke result omitted footprints")
    print(json.dumps(result, separators=(",", ":"), sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: darwin/runtime/smoke.ts
import { canonicalJson, sha256Text, wireError } from "./wire";

const HTTP_INTERNAL_SERVER_ERROR = 500;
const HEALTH_ATTEMPTS = 100;
const HEALTH_RETRY_MILLISECONDS = 50;
const LAST_HEALTH_ATTEMPT = HEALTH_ATTEMPTS - 1;
const TARGET_FOOTPRINT_BYTES = 128;
const REQUIRED_CODE_CHECK_CONSTRAINTS = 3;

const argument = (name: string): string => {
  const index = Bun.argv.indexOf(name);
  const value = index === -1 ? undefined : Bun.argv[index + 1];
  return value === undefined
    ? wireError(`${name} is required`, HTTP_INTERNAL_SERVER_ERROR)
    : value;
};

const isRecord = (value: unknown): value is Readonly<Record<string, unknown>> =>
  value !== null && typeof value === "object" && !Array.isArray(value);

const requireRecord = (
  value: unknown,
  name: string,
): Readonly<Record<string, unknown>> =>
  isRecord(value)
    ? value
    : wireError(`${name} must be an object`, HTTP_INTERNAL_SERVER_ERROR);

const endpoint = argument("--endpoint");
const path = argument("--path");
const requestPath = argument("--request");
const kind = argument("--kind");
const token = argument("--token");

const request = async (url: URL, value?: unknown): Promise<unknown> => {
  const response = await fetch(
    url,
    value === undefined
      ? undefined
      : {
        method: "POST",
        headers: {
          authorization: `Bearer ${token}`,
          "content-type": "application/json",
        },
        body: JSON.stringify(value),
      },
  );
  if (!response.ok) {
    return wireError(
      `smoke request returned HTTP ${response.status}`,
      HTTP_INTERNAL_SERVER_ERROR,
    );
  }
  return response.json();
};

const baselinePayload = (
  value: Readonly<Record<string, unknown>>,
): Readonly<Record<string, unknown>> => {
  const candidateNames = new Set([
    "candidate",
    "candidateSnapshotId",
    "benchmarkGates",
    "footprintLimits",
    "confidenceLevel",
    "bootstrapIterations",
    "multipleComparisonCount",
    "requiredConstraints",
    "evaluationPlanDigest",
  ]);
  const rawRun = value["run"];
  const dataset = value["dataset"];
  if (!isRecord(rawRun) || !isRecord(dataset)) {
    return wireError(
      "comparison fixture is invalid",
      HTTP_INTERNAL_SERVER_ERROR,
    );
  }
  const run = {
    ...Object.fromEntries(
      Object.entries(rawRun).filter(([name]) => name !== "optimizationSeed"),
    ),
    state: {
      _tag: "dataset-ready",
      datasetId: dataset["id"],
      datasetDigest: dataset["digest"],
    },
  };
  const plan = {
    ...Object.fromEntries(
      Object.entries(value).filter(([name]) => !candidateNames.has(name)),
    ),
    operation: "baseline",
    run,
    targetFootprintBytes: TARGET_FOOTPRINT_BYTES,
  };
  return {
    ...plan,
    evaluationPlanDigest: sha256Text(canonicalJson(plan)),
  };
};

for (let attempt = 0; attempt < HEALTH_ATTEMPTS; attempt += 1) {
  try {
    const health = await request(new URL("/healthz", endpoint));
    if (isRecord(health) && health["status"] === "ok") break;
  } catch {
    if (attempt === LAST_HEALTH_ATTEMPT) {
      wireError("darwin did not become healthy", HTTP_INTERNAL_SERVER_ERROR);
    }
    await Bun.sleep(HEALTH_RETRY_MILLISECONDS);
  }
}

const fixture = requireRecord(
  await Bun.file(requestPath).json(),
  "smoke fixture",
);
const payload = kind === "baseline" ? baselinePayload(fixture) : fixture;
const result = requireRecord(
  await request(new URL(path, endpoint), payload),
  "smoke result",
);

const runId = () => requireRecord(payload["run"], "run")["id"];
const candidateId = () =>
  requireRecord(payload["candidate"], "candidate")["id"];

const benchmarkInvalid = (): boolean =>
  kind === "benchmark"
  && (result["benchmarkRef"] !== payload["benchmarkRef"]
    || result["passed"] !== true);

const evaluatorInvalid = (): boolean =>
  kind === "evaluator"
  && (result["runId"] !== runId()
    || result["candidateId"] !== candidateId()
    || result["passed"] !== true);

const baselineInvalid = (): boolean =>
  kind === "baseline" && result["runId"] !== runId();

const codeCheckInvalid = (): boolean =>
  kind === "code-check"
  && (result["candidateId"] !== candidateId()
    || !Array.isArray(result["constraints"])
    || result["constraints"].length !== REQUIRED_CODE_CHECK_CONSTRAINTS);

const optimizerInvalid = (): boolean => {
  if (kind !== "optimizer") return false;
  const candidates = result["candidates"];
  if (
    result["runId"] !== runId()
    || result["paused"] !== false
    || !Array.isArray(candidates)
    || candidates.length !== 1
  ) {
    return true;
  }
  return candidates.some(
    (candidate: unknown) =>
      !(isRecord(candidate)
        && candidate["runId"] === runId()
        && typeof candidate["candidateDigest"] === "string"
        && candidate["candidateDigest"].startsWith("sha256:")),
  );
};

if (optimizerInvalid()) {
  wireError("optimizer smoke result is invalid", HTTP_INTERNAL_SERVER_ERROR);
}
if (benchmarkInvalid()) {
  wireError("benchmark smoke result is invalid", HTTP_INTERNAL_SERVER_ERROR);
}
if (evaluatorInvalid()) {
  wireError("evaluation smoke result is invalid", HTTP_INTERNAL_SERVER_ERROR);
}
if (baselineInvalid()) {
  wireError("baseline smoke result is invalid", HTTP_INTERNAL_SERVER_ERROR);
}
if (codeCheckInvalid()) {
  wireError("code-check smoke result is invalid", HTTP_INTERNAL_SERVER_ERROR);
}
